File: Day 67/Solution.java
// 125. Valid Palindrome
// A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

// Given a string s, return true if it is a palindrome, or false otherwise.
class Solution {
    public boolean isPalindrome(String s) {
        int left = 0;
        int right = s.length() - 1;
        
        while (left < right) {
            // Skip non-alphanumeric characters from left
            while (left < right && !Character.isLetterOrDigit(s.charAt(left))) {
                left++;
            }
            // Skip non-alphanumeric characters from right
            while (left < right && !Character.isLetterOrDigit(s.charAt(right))) {
                right--;
            }
            
            // Compare ignoring case
            if (Character.toLowerCase(s.charAt(left)) != Character.toLowerCase(s.charAt(right))) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }
    
    public static void main(String[] args) {
        Solution sol = new Solution();
        
        // Test case 1: "A man, a plan, a canal: Panama"
        System.out.println(sol.isPalindrome("A man, a plan, a canal: Panama")); // true
        
        // Test case 2: "race a car"
        System.out.println(sol.isPalindrome("race a car")); // false
        
        // Test case 3: " "
        System.out.println(sol.isPalindrome(" ")); // true
    }
}

File: Day 67/Solution.py
# 125. Valid Palindrome
# A phrase is a palindrome if, after converting all uppercase letters into lowercase letters and removing all non-alphanumeric characters, it reads the same forward and backward. Alphanumeric characters include letters and numbers.

# Given a string s, return true if it is a palindrome, or false otherwise.
class Solution:
    def isPalindrome(self, s: str) -> bool:
        left, right = 0, len(s) - 1
        
        while left < right:
            # Skip non-alphanumeric from left
            while left < right and not s[left].isalnum():
                left += 1
            # Skip non-alphanumeric from right
            while left < right and not s[right].isalnum():
                right -= 1
            
            # Compare ignoring case
            if s[left].lower() != s[right].lower():
                return False
            left += 1
            right -= 1
        
        return True

if __name__ == "__main__":
    sol = Solution()
    
    # Test case 1
    print(sol.isPalindrome("A man, a plan, a canal: Panama"))  # True
    
    # Test case 2
    print(sol.isPalindrome("race a car"))  # False
    
    # Test case 3
    print(sol.isPalindrome(" "))  # True
